pageserver/src/layered_repository/inmemory_layer.rs

```diff
@@ -44,7 +44,7 @@ pub struct InMemoryLayer {
 
     /// The above fields never change. The parts that do change are in 'inner',
     /// and protected by mutex.
-    inner: Mutex<InMemoryLayerInner>,
+    inner: RwLock<InMemoryLayerInner>,
```

Comment:
Do we need the RwLock because now we hold it during disk IO

Reply:
That's right. We hold it during disk I/O, but that's OK because we only hold it in read mode, and the layer has already frozen so there are no writes on it anymore. So it shouldn't block anyone.

Hmm, there is one exception to that. If this race condition in writing happens:

1. thread 1 calls put_page_version(). It calls get_layer_for_write() and returns reference to layer
2. checkpointer wakes up, and freezes the layer, and calls write_to_disk() on it.
3. Thread 1 calls layer.put_page_version(). It blocks waiting for the lock in write-mode.

Thread 1 will block until write_to_disk() has finished. Then, it will acquire the lock, see that the layer is no longer writeable, and retry the operation from the beginning. This is no different than what happens today without this PR, though.

Reply:
Right, but today the critical section doesn't wait for IO, meaning thread 1 will get the lock quicker. Regardless I don't think this is a big deal because of how rare this race probably is.

Reply:
> Right, but today the critical section doesn't wait for IO, meaning thread 1 will get the lock quicker.

True.

> Regardless I don't think this is a big deal because of how rare this race probably is.

Ok, I'll just add a comment explaining it.